pkg/ipcache/metadata.go

```diff
@@ -427,22 +427,37 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
 	}
 
 	lbls := info.ToLabels()
+	hasRemoteNode := lbls.Has(labels.LabelRemoteNode[labels.IDNameRemoteNode])
+	hasLocalHost := lbls.Has(labels.LabelHost[labels.IDNameHost])
+
+	// Add in CIDR labels where needed:
+	// - for out-of-cluster entities
+	// - for remote nodes, when policy-cidr-selects-nodes is enabled
+	// - for localhost, when policy-cidr-selects-nodes is enabled
+	if !lbls.Has(labels.LabelHealth[labels.IDNameHealth]) &&
+		!lbls.Has(labels.LabelIngress[labels.IDNameIngress]) &&
+		(!hasRemoteNode || (option.Config.PolicyCIDRSelectsNodes && hasRemoteNode)) &&
+		(!hasLocalHost || (option.Config.PolicyCIDRSelectsNodes && hasLocalHost)) {
+		cidrLabels := cidrlabels.GetCIDRLabels(prefix)
+		lbls.MergeLabels(cidrLabels)
+	}
+
+	// GetCIDRLabels added in reserved:world unconditionally; remove that
+	// from in-cluster components
 	if (lbls.Has(labels.LabelWorld[labels.IDNameWorld]) ||
 		lbls.Has(labels.LabelWorldIPv4[labels.IDNameWorldIPv4]) ||
 		lbls.Has(labels.LabelWorldIPv6[labels.IDNameWorldIPv6])) &&
-		(lbls.Has(labels.LabelRemoteNode[labels.IDNameRemoteNode]) ||
-			lbls.Has(labels.LabelHost[labels.IDNameHost])) {
+		(hasRemoteNode || hasLocalHost) {
 		// If the prefix is associated with both world and (remote-node or
-		// host), then the latter (remote-node or host) take precedence to
-		// avoid allocating a CIDR identity for an entity within the cluster.
+		// host), then the latter (remote-node or host) take precedence and
+		// world should not be applied.
 		n := lbls.Remove(labels.LabelWorld)
 		n = n.Remove(labels.LabelWorldIPv4)
 		n = n.Remove(labels.LabelWorldIPv6)
-		n = n.Remove(cidrlabels.GetCIDRLabels(prefix))
```

Comment:
I think there's a problem here, specifically in the case where the new flag is disabled.

I believe that `info.ToLabels()` above could return the full set of CIDR labels iff there is a remote node and a CIDR policy that also specifies the same prefix. Previously the code would have removed the CIDR labels because that's how the current logic handles CIDRs for remote nodes - the remote nodes are not selectable by policy. I'm concerned that removing this line here would actually cause a regression where CIDR policy will now select remote nodes even if the flag is off.

I suspect this is why the logic was previously the other way around, with the labels being first removed... then only in specific scenarios the target labels are re-added. In a sense, the `info.ToLabels()` is a relatively unvetted set of requests from a range of resources to assign specific labels to the prefix & generate a corresponding Identity. This function is the function that's enforcing the policy posture that determines how the various remote-node, CIDR labels are mixed into the Identity.

Reply:
This function is confusing. Specifically, there is an unwritten assumption that the metadata layer / InjectLabels is only called for the class of entities that _might_ have CIDR labels. Then, we remove the CIDR labels only in the case of `host` and `remote-node`. Confusingly, we then re-add them unless we see the labels `host`, `remote-node`, `health`, or `ingress`.

So that means `health` or `ingress` _might_ have CIDR labels, but they might not. Huh.

So there are four cases here

- the prefix had CIDR labels, but we removed them
- the prefix had CIDR labels, and we kept them
- the prefix did not have CIDR labels, and we didn't add them
- the prefix did not have CIDR labels, but we added them


I see your point; we might unnecessarily re-add labels in the case that they are already in lbls. I'll re-think the logic with this perspective.

Reply:
OK, I've looked at `resolveIdentity()` a bit more closely and the diff is far more minimal now.